### PR TITLE
Add InstrumentationSystem

### LINKED ISSUE #47: Consider InstrumentationSystem (?) similar to MetricsSystem et al
Libraries will often need to pull an instrument "out of thin air" in a similar way as metrics have to be pulled out sometimes.

We should offer something similar to https://github.com/apple/swift-metrics#selecting-a-metrics-backend-implementation-applications-only

Naming will evolve here but something like `InstrumentationSystem` to be consistent with Metrics and logging perhaps?
We should also think how it relates to tracing; would we allow e.g. `InstrumentationSystem.instrument` and also `InstrumentationSystem.tracer` (where a tracer is-a instrument but not all usages need a tracer etc.)

### PR DESCRIPTION
**Pull request by slashmo**
Closes #47

### CODE REVIEW

**Comment by slashmo** [comment]
@ktoso  I copied over the `Lock` parts we needed from Swift Metrics. Should we start adding license headers now?

**Comment by ktoso** [comment]
> @ktoso I copied over the Lock parts we needed from Swift Metrics.

ok good

> Should we start adding license headers now?

Yes, about time I guess.

You could copy the `validate_license_header.sh` script as well and adjust it to our project name from metrics.

File: Sources/Instrumentation/InstrumentationSystem.swift [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    ///
+    /// - Parameter instrument: The `Instrument` you want to share globally within your system.
+    /// - Warning: Do not call this method more than once. This will lead to a crash.
+    public static func bootstrap(_ instrument: Instrument) {
+        self.lock.withWriterLock {
+            precondition(
+                !self.isInitialized, """
+                InstrumentationSystem can only be initialized once per process. Consider using MultiplexInstrument if
+                you need to use multiple instruments.
+                """
```
Comment (ktoso) [inline Sources/Instrumentation/InstrumentationSystem.swift:22]: 👍

File: Sources/Instrumentation/InstrumentationSystem.swift [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        }
+    }
+
+    /// Returns the globally configured `Instrument`. Defaults to a no-op `Instrument` if `boostrap` wasn't called before.
+    public static var instrument: Instrument {
+        self.lock.withReaderLock { self._instrument }
+    }
+}
+
+private final class NOOPInstrument: Instrument {
```
Comment (ktoso) [inline Sources/Instrumentation/InstrumentationSystem.swift:35]: 👍 maybe in its own file though?

File: Sources/Instrumentation/Locks.swift [hunk 3]
```diff
@@ -0,0 +1,89 @@
+#if os(macOS) || os(iOS) || os(tvOS) || os(watchOS)
```
Comment (ktoso) [inline Sources/Instrumentation/Locks.swift:1]: copy license header as well from where these files are taken? (metrics or logging)

**Review by ktoso — APPROVED** [review]
LGTM :) Maybe split out the NOOP impl into its own file.

Follow up with license headers?

File: Sources/Instrumentation/InstrumentationSystem.swift (hunk 2 above)
Comment (slashmo) [reply Sources/Instrumentation/InstrumentationSystem.swift:35]: Then we'd have to make it internal. Would that be worth it?

Comment (slashmo) [reply Sources/Instrumentation/InstrumentationSystem.swift:35]: Are there other uses of such a `NOOPInstrument`?

**Pull request merged**